karapace/coordinator/schema_coordinator.py

```diff
@@ -0,0 +1,935 @@
+            LOG.error("Heartbeat send request failed: %s. Will retry.", err)
+            return False
+        error_type = Errors.for_code(resp.error_code)
+        if error_type is Errors.NoError:
+            LOG.debug("Received successful heartbeat response for group %s", self.group_id)
+            return True
+        if error_type in (Errors.GroupCoordinatorNotAvailableError, Errors.NotCoordinatorForGroupError):
+            LOG.warning(
+                "Heartbeat failed for group %s: coordinator (node %s) is either not started or not valid",
+                self.group_id,
+                self.coordinator_id,
+            )
+            self.coordinator_dead()
+        elif error_type is Errors.RebalanceInProgressError:
+            LOG.warning("Heartbeat failed for group %s because it is rebalancing", self.group_id)
+            self.request_rejoin()
+            # it is valid to continue heartbeating while the group is
+            # rebalancing. This ensures that the coordinator keeps the
+            # member in the group for as long as the duration of the
+            # rebalance timeout. If we stop sending heartbeats,
+            # however, then the session timeout may expire before we
+            # can rejoin.
+            return True
+        elif error_type is Errors.IllegalGenerationError:
+            LOG.warning("Heartbeat failed for group %s: generation id is not current.", self.group_id)
+            self.reset_generation()
+        elif error_type is Errors.UnknownMemberIdError:
+            LOG.warning("Heartbeat failed: local member_id was not recognized; resetting and re-joining group")
+            self.reset_generation()
+        elif error_type is Errors.GroupAuthorizationFailedError:
+            raise error_type(self.group_id)
+        else:
+            kafka_error = Errors.KafkaError(f"Unexpected exception in heartbeat task: {error_type()!r}")
+            LOG.error("Heartbeat failed: %r", kafka_error)
+            raise kafka_error
+        return False
+
+    async def _do_rejoin_group(self) -> bool:
+        rebalance = SchemaCoordinatorGroupRebalance(
+            self,
+            self.group_id,
+            self.coordinator_id,
+            self._session_timeout_ms,
+            self._retry_backoff_ms,
+        )
+        assignment = await rebalance.perform_group_join()
+
+        if assignment is None:
+            # wait backoff and try again
+            await asyncio.sleep(self._retry_backoff_ms / 1000)
+            return False
+
+        protocol, member_assignment_bytes = assignment
+        await self._on_join_complete(self.generation, self.member_id, protocol, member_assignment_bytes)
+        return True
+
+
+class SchemaCoordinatorGroupRebalance:
+    """ An adapter, that encapsulates rebalance logic
+        On how to handle cases read in https://cwiki.apache.org/confluence/\
+            display/KAFKA/Kafka+Client-side+Assignment+Proposal
+    """
```

Comment:
need to take a while for reading the KIP and read the implementation, doing it in a second look at this pr